benchmarking/process_json_to_csv.jl

```diff
@@ -0,0 +1,171 @@
+# The layout of the results is specified by a JSON file of the following format:
+#
+# {"datasets": [
+#   {"name": "name1", "logs_directory": "path1"},
+#   {"name": "name2", "logs_directory": "path2"}
+#  ] }
+#
+# For example, suppose /tmp/layout.json contains the following content:
+# {"datasets": [
+#   {"name": "pdhg,1e-8", "logs_directory": "/tmp/pdhg_1e-8"},
+#   {"name": "pdhg,1e-4", "logs_directory": "/tmp/pdhg_1e-4"},
+#   {"name": "scs,1e-8", "logs_directory": "/tmp/scs_1e-8"},
+#   {"name": "scs,1e-4", "logs_directory": "/tmp/scs_1e-4"}
+#  ] }
+#
+# Then:
+# $ julia --project=. process_json_to_csv.jl /tmp/layout.json /tmp/dataset.csv
+#
+# will generate a result file in /tmp/dataset.csv containing the summary of each
+# solve in all of four directories, with columns identifying the corresponding
+# configuration.
+
+import CSV
+import DataFrames
+import Glob
+import JSON3
+import StructArrays
+import StructTypes
+
+import FirstOrderLp
+
+const DataFrame = DataFrames.DataFrame
+
+mutable struct CsvRow
+  experiment_name::String
+  instance_name::String
+
+  # These fields correpond to those in ConvergenceInformation.
+  primal_objective::Float64
+  dual_objective::Float64
+  relative_optimality_gap::Float64
+  l2_primal_residual::Float64
+  l_inf_primal_residual::Float64
+  l2_dual_residual::Float64
+  l_inf_dual_residual::Float64
+  relative_l2_primal_residual::Float64
+  relative_l_inf_primal_residual::Float64
+  relative_l2_dual_residual::Float64
+  relative_l_inf_dual_residual::Float64
+  l_inf_primal_variable::Float64
+  l2_primal_variable::Float64
+  l_inf_dual_variable::Float64
+
+  termination_reason::FirstOrderLp.TerminationReason
+  iteration_count::Int32
+  cumulative_kkt_matrix_passes::Float64
+  solve_time_sec::Float64
+end
+
+function CsvRow()
+  return CsvRow(
+    "",  # experiment_name
+    "",  # instance_name
+    NaN,  # primal_objective
+    NaN,  # dual_objective
+    NaN,  # relative_optimality_gap
+    NaN,  # l2_primal_residual
+    NaN,  # l_inf_primal_residual
+    NaN,  # l2_dual_residual
+    NaN,  # l_inf_dual_residual
+    NaN,  # relative_l2_primal_residual
+    NaN,  # relative_l_inf_primal_residual
+    NaN,  # relative_l2_dual_residual
+    NaN,  # relative_l_inf_dual_residual
+    NaN,  # l_inf_primal_variable
+    NaN,  # l2_primal_variable
+    NaN,  # l_inf_dual_variable
+    FirstOrderLp.TERMINATION_REASON_UNSPECIFIED,  # termination_reason
+    0,  # iteration_count
+    NaN,  # cumulative_kkt_matrix_passes
+    NaN,  # solve_time_sec
+  )
+end
+
+function set_matching_fields(to, from)
+  for field_name in intersect(fieldnames(typeof(to)), fieldnames(typeof(from)))
+    setfield!(to, field_name, getfield(from, field_name))
+  end
+end
+
+function solve_log_to_csv_row(
+  log::FirstOrderLp.SolveLog,
+  experiment_name::String,
+)::CsvRow
+  row = CsvRow()
+  set_matching_fields(row, log)
+  row.experiment_name = experiment_name
```

Comment:
tiny tweak: at least to me it seems more natural to set the experiment_name before set_matching_fields(row, log) - the experiment name is the major "key" of the record (the instance_name, from the log, is the minor "key").

Reply:
Done